webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultStudentGroupBatchService.java

```diff
@@ -0,0 +1,95 @@
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+@Service
+class DefaultStudentGroupBatchService implements StudentGroupBatchService {
+
+    private final StudentGroupBatchRepository repository;
+    private final ArchiveService archiveService;
+
+    @Autowired
+    public DefaultStudentGroupBatchService(@NotNull final StudentGroupBatchRepository repository,
+                                           final ArchiveService archiveService) {
+        this.repository = repository;
+        this.archiveService = archiveService;
+    }
+
+    @Override
+    public Iterable<StudentGroupBatch> findAllStudentGroupBatches(final User user) {
+        return repository.findByCreator(user.getUsername());
+    }
+
+    @Override
+    public StudentGroupBatch upload(final User user, final MultipartFile file) {
+
+        final StudentGroupBatch.Builder studentGroupBatchBuilder = StudentGroupBatch.builder().creator(user.getUsername());
+
+        try {
+            final String digest = DigestUtils.md5Hex(validate(file.getBytes())).toUpperCase();
+            final Properties properties = new Properties();
+            properties.setProperty("content-type", file.getContentType());
+            properties.setProperty("username", user.getUsername());
+            properties.setProperty("size", Long.toString(file.getSize()));
+
+            //TODO: check if the digest already exists
+
+            final String location = new LocationStrategy.GroupUploadContentLocationStrategy().location(digest);
+            archiveService.writeResource(location, file.getInputStream(), properties);
+
+            studentGroupBatchBuilder
+                    .digest(digest)
+                    .status(ImportStatus.ACCEPTED);
+
+        } catch (final Exception ex) {
+            studentGroupBatchBuilder
+                    .digest("") //TODO: should digest be optional?
+                    .message(ex.getMessage())
+                    .status(ImportStatus.BAD_DATA);
+        }
+        return repository.create(studentGroupBatchBuilder.build());
+
+    }
+
+    // TODO: add more validation ?
+    private byte[] validate(final byte[] input) {
+        try (final Reader reader = new InputStreamReader(new ByteArrayInputStream(input), Charset.forName("utf8"))) {
+            final Iterator<CSVRecord> recordIterator = CSVFormat.RFC4180.withFirstRecordAsHeader().parse(reader).iterator();
+            if (!recordIterator.hasNext()) throw new IllegalArgumentException("Empty file");
+
+            final Set<String> headers = recordIterator.next().toMap().keySet().stream().map((v) -> v.toLowerCase()).collect(Collectors.toSet());
+            if (!headers.containsAll(newArrayList("school_natural_id", "subject_code", "group_name", "group_user_login", "school_year", "student_ssid"))) {
+                throw new IllegalArgumentException("Missing headers, found [" + headers + "]");
```

Comment:
yes, these are different that what I had but i like these except group_user_login,
maybe group_admin_login ?